README.md
# React Autocomplete

Accessible, extensible, Autocomplete for React.js.

Things you should know about this fork:

1. It's a fork from `https://github.com/reactjs/react-autocomplete` so that we maintain its development
2. Looking for contributors
3. I am trying to revive the original repo
4. I am udpdating dependencies, code, adding and improving a11y, etc. Things might break. I should have done beta releases, but now is too late. I will then use the next major version to tell when things are more stable. For now, use it, find bugs, raise them here and or open pull requets!

```jsx
<Autocomplete
  getItemValue={(item) => item.id}
  suggestionsMenuId="input-name-suggestions"
  items={[
    { id: "apple", label: "apple" },
    { id: "banana", label: "banana" },
    { id: "pear", label: "pear" },
  ]}
  renderItem={(item, isHighlighted) => (
    <div
      key={item.id}
      role="option"
      style={{ background: isHighlighted ? "lightgray" : "white" }}
    >
      {item.label}
    </div>
  )}
  value={this.state.value}
  onChange={(e) => this.setState({ value: e.target.value })}
  onSelect={(value) => this.setState({ value })}
/>
```

Check out [more examples](https://reactcommunity.org/react-autocomplete/) and get stuck right in with the [online editor](http://jsbin.com/mipesawapi/edit?js,output).

## Install

### npm

```bash
npm install --save @learnwithgurpreet/react-autocomplete
```

### yarn

```bash
yarn add @learnwithgurpreet/react-autocomplete
```

## Accessibility

The main goal of this library is to be accessible to all users. Screen reader and keyboard only users specially. In order to do that, it controls some parts of the markup, and exposes some others for your adaptations.

Particularly, it exposes functions that allow you to decide for yourself the DOM of:

- Menu
- Item
- Input

In order to make the component accessible, it sticks to the standard defined by WAI-ARIA 1.1. Examples of basic markup can be found [here](https://www.w3.org/TR/wai-aria-practices-1.1/examples/combobox/aria1.1pattern/listbox-combo.html).

#### Patterns that you can use easily with this lib

1. **List autocomplete with manual selection:** When the popup is triggered, it presents suggested values that complete or logically correspond to the characters typed in the textbox. The character string the user has typed will become the value of the textbox unless the user selects a value in the popup.
   How? Set these props to false:

```
selectOnBlur={false}
autoHighlight={false}
```

2. **List autocomplete with automatic selection:** When the popup is triggered, it presents suggested values that complete or logically correspond to the characters typed in the textbox, and the first suggestion is automatically highlighted as selected. The automatically selected suggestion becomes the value of the textbox when the combobox loses focus unless the user chooses a different suggestion or changes the character string in the textbox.
   How? Set these props to true:

```
selectOnBlur={true}
autoHighlight={true}
```

And to understand the comobobox pattern, the best place to start is [to read the spec](https://www.w3.org/TR/wai-aria-practices-1.1/#combobox).

Having said that, in the demo page, you will see 4 examples. Some follow the pattern List autocomplete with manual selection, others follow the pattern List autocomplete with automatic selection.

### Things to mind when having custom implementations of any of the `render` properties.

#### RenderItem

Appart from all other attributes to have a valid markup. Note: This library will add automatically an ID to the item, will be the concatenation of the `suggestionsMenuId` with `item-{itemIndex}`; It will be used to define the active descendant.

why do we need active-descendant? because it will manage the focus to the screen reader.

- To `renderItem` add for example:

```js
 renderItem={(item, isHighlighted) => (
    <div
      role="option"
      aria-selected={isHighlighted}
      key={item.abbr}
    >{item.name}</div>
  )}
```

#### RenderMenu

Be sure to add the `id` and the `role=listbox`

```jsx
id = { suggestionsMenuId };
role = "listbox";
```

#### RenderInput

First of all, you don't want the browser autcomplete (like Chrome's one) to appear on top of your suggestions. For that, switch of the autocomplete attribute.

Second, the standard does not show how to add instructions, but it is a nice thing to add, so that the screen reader user knows that they are not only in a text field, but they also know they are in a combobox, that gives them the chance to select suggestions from below. To do that, add the `aria-describedby` pointing to an element in the DOM with the instructions. (this is shown in the examples, you can test it there)

```jsx
  autoComplete="something-that-is-not-off"
  aria-describedby="the-id-to-instructions-if-any"
```

Example of instructions DOM element:

```jsx
<span id="init-Instructions" className="sr-only">
  When autocomplete results are available use up and down arrows to review and
  enter to select. Touch device users, explore by touch or with swipe gestures.
</span>
```

**Note:** Don't follow these guides blindly. Test them in a real screen reader. Does not take so much. [Here some cheat sheets](https://dequeuniversity.com/screenreaders/)

## API

### Props

#### `getItemValue: Function`

Arguments: `item: Any`

Used to read the display value from each entry in `items`.

#### `suggestionsMenuId`: string

Default value: ''

Will be used in aria-owns of the input field, and the id of the suggestions menu to let
screen readers know where to find the suggestions of the autocomplete

#### `items: Array`

The items to display in the dropdown menu

#### `renderItem: Function`

Arguments: `item: Any, isHighlighted: Boolean, styles: Object`

Invoked for each entry in `items` that also passes `shouldItemRender` to
generate the render tree for each item in the dropdown menu. `styles` is
an optional set of styles that can be applied to improve the look/feel
of the items in the dropdown menu.

#### `numberOfResultsAvailableCopy`: string (optional)

Default value: 'Autocomplete results are available below.'

A string that will be added to the notification that tells the screen reader user the amount of results available.

#### `autoHighlight: Boolean` (optional)

Default value: `true`

Whether or not to automatically highlight the top match in the dropdown
menu.

#### `openOnFocus: Boolean` (optional)

Default value: `false`

If set to true, the suggestion list will be shown on input focus.

#### `inputProps: Object` (optional)

Default value: `{}`

Props passed to `props.renderInput`. By default these props will be
applied to the `<input />` element rendered by `Autocomplete`, unless you
have specified a custom value for `props.renderInput`. Any properties
supported by `HTMLInputElement` can be specified, apart from the
following which are set by `Autocomplete`: value, autoComplete, role,
aria-autocomplete. `inputProps` is commonly used for (but not limited to)
placeholder, event handlers (onFocus, onBlur, etc.), autoFocus, etc..

#### `isItemSelectable: Function` (optional)

Default value: `function() { return true }`

Arguments: `item: Any`

Invoked when attempting to select an item. The return value is used to
determine whether the item should be selectable or not.
By default all items are selectable.

#### `menuStyle: Object` (optional)

Default value:

```jsx
{
  borderRadius: '3px',
  boxShadow: '0 2px 12px rgba(0, 0, 0, 0.1)',
  background: 'rgba(255, 255, 255, 0.9)',
  padding: '2px 0',
  fontSize: '90%',
  position: 'fixed',
  overflow: 'auto',
  maxHeight: '50%', // TODO: don't cheat, let it flow to the bottom
}
```

Styles that are applied to the dropdown menu in the default `renderMenu`
implementation. If you override `renderMenu` and you want to use
`menuStyle` you must manually apply them (`this.props.menuStyle`).

#### `onChange: Function` (optional)

Default value: `function() {}`

Arguments: `event: Event, value: String`

Invoked every time the user changes the input's value.

#### `onMenuVisibilityChange: Function` (optional)

Default value: `function() {}`

Arguments: `isOpen: Boolean`

Invoked every time the dropdown menu's visibility changes (i.e. every
time it is displayed/hidden).

#### `onSelect: Function` (optional)

Default value: `function() {}`

Arguments: `value: String, item: Any`

Invoked when the user selects an item from the dropdown menu.

#### `renderInput: Function` (optional)

Default value:

```jsx
function(props) {
  return <input {...props} />
}
```

Arguments: `props: Object`

Invoked to generate the input element. The `props` argument is the result
of merging `props.inputProps` with a selection of props that are required
both for functionality and accessibility. At the very least you need to
apply `props.ref` and all `props.on<event>` event handlers. Failing to do
this will cause `Autocomplete` to behave unexpectedly.

#### `renderMenu: Function` (optional)


CHANGELOG.md
- added build command in github actions
- changed script to publish the component on npm
- added github action
- 3.0.0
- Update README
- 3.0.0
- 2.0.0
- 1.0.2
- 1.0.1
- Merge pull request #2 from learnwithgurpreet/feat/add-change-log
- faet(chore): added script to genereate change log
- Merge pull request #1 from learnwithgurpreet/fix/a11y-aria-activedescendant
- made changes for new npm module setup
- fix(A11y): Fixed aria-activedescendant related problem
- Add openOnFocus to readme
- 2.3.0
- Update changelog for version 2.2.0
- 2.2.0
- feat: Add openOnFocus feature
- 2.1.0
- Merge pull request #7 from DanielaValero/fix/remove-open-and-fix-imperative-api
- add release notes v2.1.0
- feat: open menu only when there are values and not on focus
- build: Update to react 16.9.0
- build: remove menu example from localhost
- docs: Remove example of outside management of menu visibility
- 2.0.0
- Update README
- docs: Update changelog of version 2 and documentation
- fix: Call onBlur and onFocus in imperativeAPI and remove openProp
- Update changelog to v1.8.12
- 1.8.11
- Merge pull request #6 from DanielaValero/fix/resolve-scrolling-issue
- fix: Resolve issues with scrolling to item and add unit test
- build: update some of the dependencies
- 1.8.11-0
- 1.8.10-0
- docs: add release notes for 1.8.9-1
- build: Point to built compiled package for npm
- 1.8.9-0
- build: dont add lib to npm to try to get things updated
- 1.8.8
- Update README
- docs: Add release notes for version 1.8.8
- docs: Add notes in readme about fork state
- Merge pull request #5 from DanielaValero/feat/stabilization-and-a11y
- feat: dont show hint of results below if empty text
- test: Select based on aria roles instead of div and positions in dom
- feat: extend screen readers functionality
- fix: Append properly the events to Automplete for the imperative API
- docs: Update documentation of how to make accessible to screen readers
- 1.8.7
- Update README
- changelog: Add release notes for 1.8.7"
- Merge pull request #4 from DanielaValero/feat/3-screen-reader
- feat: Update all examples with a11y features
- chore: let example closed
- build: Install only lodas memoize
- refactor: Remove setting menu positions did not see why is required
- feat(a11y): Make component usable with screen readers
- refactor: Start to migrate deprecated standards
- feat: Autohighligh entered text after component update
- chore: destructre props and state
- chore: beautify code
- chore: access state and props from class rather than args
- docs: Add migration guide
- feat: Drop support of Node < 6
- changelog: fix release version
- 1.8.6
- 1.8.5
- changelog: Add release notes 1.8.5
- Merge pull request #2 from DanielaValero/build/update-deps
- build: Update deps with full migration of deprecated feats like ref strings
- build: fix build command with new requirements of dependencies
- build: Start to Update all dependencies
- 1.8.4
- Update README
- build: update links and stuff
- 1.8.3
- Update README
- test making a release
- docs: rename package and update docs
- docs: update examples with new prop
- feat: Make accessible to screen readers and allow autocomplete to be set
- Update README
- 1.8.1
- Update README
- Update CHANGELOG with 1.8.1 release
- Merge pull request #308 from mattraykowski/mobile-blur-fix
- Merge pull request #315 from rssteffey/fix-arrow-up
- Fix ArrowUp to properly skip over unselectable items
- 1.8.0
- Update README
- Update CHANGELOG with 1.8.0 release
- Update Custom Menu example to use isItemSelectable
- Add props.isItemSelectable
- Update README
- 1.7.3
- Update README
- Update CHANGELOG with 1.7.3 release
- Release focus-lock when user uses Enter to close menu
- Fixed #307: ignore blur on touch start.
- 1.7.2
- Update README
- Update CHANGELOG with 1.7.2 release
- Merge pull request #280 from bell-steven/master
- add react 16 as a peer dependency
- 1.7.1
- Update CHANGELOG with 1.7.1 release
- Prevent item selection when keyCode != 13
- Fix incorrect version number in latest release notes
- Update README
- 1.7.0
- Update README
- Update CHANGELOG with 1.6.0 and 1.7.0 releases
- Add props.selectOnBlur
- Remove `styles` from utils
- 1.6.0
- Update README
- Add props.renderInput
- Enable snapshotting of enzyme wrapper objects
- Update README
- Add link to JSBin example in README
- Update README
- Add example uses to inputProps documentation
- Update README
- 1.5.10
- Update CHANGELOG with 1.5.10 release

lib/Autocomplete.js
const React = require('react')
const memoize = require('lodash.memoize');
const PropTypes = require('prop-types')
const scrollIntoView = require('dom-scroll-into-view')

const IMPERATIVE_API = [
  'blur',
  'checkValidity',
  'click',
  'focus',
  'select',
  'setCustomValidity',
  'setSelectionRange',
  'setRangeText',
]

function getScrollOffset() {
  return {
    x: (window.pageXOffset !== undefined) ?
      window.pageXOffset :
      (document.documentElement || document.body.parentNode || document.body).scrollLeft,
    y: (window.pageYOffset !== undefined) ?
      window.pageYOffset :
      (document.documentElement || document.body.parentNode || document.body).scrollTop,
  }
}

class Autocomplete extends React.Component {

  static propTypes = {
    /**
     * The items to display in the dropdown menu
     */
    items: PropTypes.array.isRequired,
    /**
     * The value to display in the input field
     */
    value: PropTypes.any,
    /**
     * Arguments: `event: Event, value: String`
     *
     * Invoked every time the user changes the input's value.
     */
    onChange: PropTypes.func,
    /**
     * Arguments: `value: String, item: Any`
     *
     * Invoked when the user selects an item from the dropdown menu.
     */
    onSelect: PropTypes.func,
    /**
     * Arguments: `item: Any, value: String`
     *
     * Invoked for each entry in `items` and its return value is used to
     * determine whether or not it should be displayed in the dropdown menu.
     * By default all items are always rendered.
     */
    shouldItemRender: PropTypes.func,
    /**
     * Arguments: `item: Any`
     *
     * Invoked when attempting to select an item. The return value is used to
     * determine whether the item should be selectable or not.
     * By default all items are selectable.
     */
    isItemSelectable: PropTypes.func,
    /**
     * Arguments: `itemA: Any, itemB: Any, value: String`
     *
     * The function which is used to sort `items` before display.
     */
    sortItems: PropTypes.func,
    /**
     * Arguments: `item: Any`
     *
     * Used to read the display value from each entry in `items`.
     */
    getItemValue: PropTypes.func.isRequired,
    /**
     * Arguments: `item: Any, isHighlighted: Boolean, styles: Object`
     *
     * Invoked for each entry in `items` that also passes `shouldItemRender` to
     * generate the render tree for each item in the dropdown menu. `styles` is
     * an optional set of styles that can be applied to improve the look/feel
     * of the items in the dropdown menu.
     */
    renderItem: PropTypes.func.isRequired,
    /**
     * Arguments: `items: Array<Any>, value: String, styles: Object`
     *
     * Invoked to generate the render tree for the dropdown menu. Ensure the
     * returned tree includes every entry in `items` or else the highlight order
     * and keyboard navigation logic will break. `styles` will contain
     * { top, left, minWidth } which are the coordinates of the top-left corner
     * and the width of the dropdown menu.
     */
    renderMenu: PropTypes.func,
    /**
     * Styles that are applied to the dropdown menu in the default `renderMenu`
     * implementation. If you override `renderMenu` and you want to use
     * `menuStyle` you must manually apply them (`this.props.menuStyle`).
     */
    menuStyle: PropTypes.object,
    /**
     * Arguments: `props: Object`
     *
     * Invoked to generate the input element. The `props` argument is the result
     * of merging `props.inputProps` with a selection of props that are required
     * both for functionality and accessibility. At the very least you need to
     * apply `props.ref` and all `props.on<event>` event handlers. Failing to do
     * this will cause `Autocomplete` to behave unexpectedly.
     */
    renderInput: PropTypes.func,
    /**
     * Props passed to `props.renderInput`. By default these props will be
     * applied to the `<input />` element rendered by `Autocomplete`, unless you
     * have specified a custom value for `props.renderInput`. Any properties
     * supported by `HTMLInputElement` can be specified, apart from the
     * following which are set by `Autocomplete`: value, autoComplete, role,
     * aria-autocomplete. `inputProps` is commonly used for (but not limited to)
     * placeholder, event handlers (onFocus, onBlur, etc.), autoFocus, etc..
     */
    inputProps: PropTypes.object,
    /**
     * Props that are applied to the element which wraps the `<input />` and
     * dropdown menu elements rendered by `Autocomplete`.
     */
    wrapperProps: PropTypes.object,
    /**
     * This is a shorthand for `wrapperProps={{ style: <your styles> }}`.
     * Note that `wrapperStyle` is applied before `wrapperProps`, so the latter
     * will win if it contains a `style` entry.
     */
    wrapperStyle: PropTypes.object,
    /**
     * Whether or not to automatically highlight the top match in the dropdown
     * menu.
     */
    autoHighlight: PropTypes.bool,
    /**
     * Whether or not to automatically select the highlighted item when the
     * `<input>` loses focus.
     */
    selectOnBlur: PropTypes.bool,
    /**
     * Opens the menu on focus even when the input field is empty.
     */
    openOnFocus: PropTypes.bool,
    /**
     * Arguments: `isOpen: Boolean`
     *
     * Invoked every time the dropdown menu's visibility changes (i.e. every
     * time it is displayed/hidden).
     */
    onMenuVisibilityChange: PropTypes.func,
    debug: PropTypes.bool,
    /** Set the id of the menu of suggestions.
     * will be used to link the input with it, so
     * screen readers find it
     */
    suggestionsMenuId: PropTypes.string.isRequired,
    /** String. Label to be used to tell screen
     * reader users how many results are available */
    numberOfResultsAvailableCopy: PropTypes.string,
  }

  static defaultProps = {
    value: '',
    wrapperProps: {},
    wrapperStyle: {
      display: 'inline-block'
    },
    inputProps: {},
    renderInput(props) {
      return <input {...props} />
    },
    onChange() {},
    onSelect() {},
    isItemSelectable() {
      return true
    },
    renderMenu(items, suggestionsMenuId, value, style) {
      return <div id={suggestionsMenuId} role="listbox" style={{ ...style }}>{items}</div>
    },
    menuStyle: {
      borderRadius: '3px',
      boxShadow: '0 2px 12px rgba(0, 0, 0, 0.1)',
      background: 'rgba(255, 255, 255, 0.9)',
      padding: '2px 0',
      fontSize: '90%',
      position: 'fixed',
      overflow: 'auto',
      maxHeight: '50%',
    },
    autoHighlight: true,
    selectOnBlur: false,
    openOnFocus: false,
    onMenuVisibilityChange() {},
    numberOfResultsAvailableCopy: 'Autocomplete results are available below.'
  }

  constructor(props) {
    super(props)
    this.state = {
      isOpen: false,
      highlightedIndex: null,
      filteredItems: props.items,
    }
    this._debugStates = []
    this.references = {
      input: React.createRef(),
      menu: {
        current: null
      },
    }
    this._ignoreBlur = false
    this._ignoreFocus = false
    this._scrollOffset = null
    this._scrollTimer = null
  }

  getMemoizedFilteredItems = memoize(({ items, value })=> this.getFilteredItemsToMemoize(items, value));

  componentWillUnmount() {
    clearTimeout(this._scrollTimer)
    this._scrollTimer = null
  }

  UNSAFE_componentWillReceiveProps() {
    if (this.state.highlightedIndex !== null) {
      const highlightedIndex = this.ensureHighlightedIndex();
      this.setState({
        ...highlightedIndex
      });
    }
  }

  componentDidMount() {
    this.exposeAPI();
  }

  componentDidUpdate(prevProps, prevState) {
    const {
      isOpen,
      highlightedIndex,
    } = this.state;

    const {
      onMenuVisibilityChange,
      autoHighlight,
    } = this.props;

    this.maybeScrollItemIntoView()

    if (prevState.isOpen !== isOpen) {
      onMenuVisibilityChange(isOpen)
    }

    if (autoHighlight && (prevProps.value !== this.props.value || highlightedIndex === null)) {

      this.maybeAutoCompleteText();
   }
  }

  exposeAPI = () => {
    IMPERATIVE_API.forEach(event => {
      const enventHandler = this.references.input.current && this.references.input.current[event];
      if (enventHandler) {
        // this[event] = () => {}
        if(event === 'focus'){
          this[event] = this.handleInputFocus;
          return
        }

        if(event === 'blur'){
          this[event] = this.handleInputBlur
          return
        }

        this[event] = enventHandler

      }
      return this[event];
    });
  }

  maybeScrollItemIntoView = () => {
    const {
      highlightedIndex,
      isOpen,
    } = this.state;

    const currentItem = this.references[`item-${highlightedIndex}`] && this.references[`item-${highlightedIndex}`].current;
    const itemNode = highlightedIndex !== null ? currentItem : null;
    const menuNode = this.references.menu.current;

    if (isOpen && !!itemNode && !!menuNode) {
      scrollIntoView(
        itemNode,
        menuNode, {
          onlyScrollIfNeeded: true
        }
      )
    }
  }

  handleKeyDown = (event) => {
    const {
      isOpen,
    } = this.state;

    if (this.keyDownHandlers[event.key]){
      this.keyDownHandlers[event.key].call(this, event);
    } else if (!isOpen) {
      this.setState({
        isOpen: true
      })
    }
  }

  handleChange = (event) => {
    const {
      onChange,
      openOnFocus,
    } = this.props;

    if(!openOnFocus) {
      this.setState({
        isOpen: event.target.value !== ''
      })
    }

    onChange(event, event.target.value)
  }

  keyDownHandlers = {
    ArrowDown(event) {
      event.preventDefault()

      const {
        isItemSelectable,
        items,
        value,
      } = this.props;

      const {
        highlightedIndex
      } = this.state;

     // const items = this.getFilteredItems()
     const memoizedItems = this.getMemoizedFilteredItems({items, value});

      if (!memoizedItems.length) return

      let index = highlightedIndex === null ? -1 : highlightedIndex

      for (let i = 0; i < memoizedItems.length; i++) {
        const p = (index + i + 1) % memoizedItems.length

        if (isItemSelectable(memoizedItems[p])) {
          index = p
          break
        }
      }
      if (index > -1 && index !== highlightedIndex) {
        this.setState({
          highlightedIndex: index,
          isOpen: true,
        })
      }
    },

    ArrowUp(event) {
      event.preventDefault()
      const {
        isItemSelectable,
        items,
        value,
      } = this.props;

      const memoizedItems = this.getMemoizedFilteredItems({items, value});


      if (!memoizedItems.length) return

      const {
        highlightedIndex
      } = this.state

      let index = highlightedIndex === null ? memoizedItems.length : highlightedIndex

      for (let i = 0; i < memoizedItems.length; i++) {
        const p = (index - (1 + i) + memoizedItems.length) % memoizedItems.length
        if (isItemSelectable(memoizedItems[p])) {
          index = p
          break
        }
      }

      if (index !== memoizedItems.length) {
        this.setState({
          highlightedIndex: index,
          isOpen: true,
        })
      }
    },

    Enter(event) {

      const {
        highlightedIndex,
        isOpen,
      } = this.state;

      const {
        getItemValue,
        onSelect,
      } = this.props;

      // Key code 229 is used for selecting items from character selectors (Pinyin, Kana, etc)
      if (event.keyCode !== 13) return
      // In case the user is currently hovering over the menu
      this.setIgnoreBlur(false)
      if (!isOpen) {
        // menu is closed so there is no selection to accept -> do nothing
        return
      } else if (highlightedIndex == null) {
        // input has focus but no menu item is selected + enter is hit -> close the menu, highlight whatever's in input
        this.setState({
          isOpen: false
        }, () => {
          this.references.input.current.select()
        })
      } else {
        // text entered + menu item has been highlighted + enter is hit -> update value to that of selected menu item, close the menu
        event.preventDefault()
        const item = this.getFilteredItems()[highlightedIndex]
        const value = getItemValue(item)

        this.setState({
          isOpen: false,
          highlightedIndex: null
        }, () => {
          //this.refs.input.focus() // TODO: file issue
          this.references.input.current.setSelectionRange(
            value.length,
            value.length
          )
          onSelect(value, item)
        })
      }
    },

    Escape() {
      // In case the user is currently hovering over the menu
      this.setIgnoreBlur(false)
      this.setState({
        highlightedIndex: null,
        isOpen: false
      })
    },

    Tab() {
      // In case the user is currently hovering over the menu
      this.setIgnoreBlur(false)
    },
  }

  getFilteredItemsToMemoize = (items, value) => {
    const {
      shouldItemRender,
      sortItems,
    } = this.props;

    let filteredItems = items;

    if (shouldItemRender) {
      filteredItems = items.filter((item) => (
        shouldItemRender(item, value)
      ))
    }

    if (sortItems) {
      filteredItems.sort((a, b) => (
        sortItems(a, b, value)
      ))
    }
    return filteredItems
  }

  getFilteredItems = () => {
    const {
      items,
      shouldItemRender,
      sortItems,
      value,
    } = this.props;

    let filteredItems = items;

    if (shouldItemRender) {
      filteredItems = items.filter((item) => (
        shouldItemRender(item, value)
      ))
    }

    if (sortItems) {
      filteredItems.sort((a, b) => (
        sortItems(a, b, value)
      ))
    }

    return filteredItems
  }

  maybeAutoCompleteText = () => {
    const {
      value,
      getItemValue,
      isItemSelectable,
    } = this.props
    const {
      highlightedIndex
    } = this.state;

    let newHighlightedIndex = null;

    let index = highlightedIndex === null ? 0 : highlightedIndex
    let items = this.getFilteredItems()

    for (let i = 0; i < items.length; i++) {
      if (isItemSelectable(items[index]))
        break
      index = (index + 1) % items.length
    }

    const matchedItem = items[index] && isItemSelectable(items[index]) ? items[index] : null

    if (value !== '' && matchedItem) {
      const itemValue = getItemValue(matchedItem);
      const itemValueDoesMatch = (itemValue.toLowerCase().indexOf(value.toLowerCase()) === 0);

      if (itemValueDoesMatch) {
        newHighlightedIndex = index;
      }
    } else {
      newHighlightedIndex = null;
    }

     if (newHighlightedIndex !== highlightedIndex) {
        this.setState({
          highlightedIndex: newHighlightedIndex
        });
      }
  }

  ensureHighlightedIndex = () => {
    const {
      highlightedIndex
    } = this.state;

    if (highlightedIndex >= this.getFilteredItems().length) {
      return {
        highlightedIndex: null
      }
    }
  }

  highlightItemFromMouse = (index) => {
    this.setState({
      highlightedIndex: index
    })
  }

  selectItemFromMouse = (item) => {
    const value = this.props.getItemValue(item)
    // The menu will de-render before a mouseLeave event
    // happens. Clear the flag to release control over focus
    this.setIgnoreBlur(false)
    this.setState({
      isOpen: false,
      highlightedIndex: null
    }, () => {
      this.props.onSelect(value, item)
    })
  }

  setIgnoreBlur = (ignore) => {
    this._ignoreBlur = ignore
  }

  renderItem = (memoizedItems) => {
    const {
      highlightedIndex,
    } = this.state;

    const {
      renderItem,
      isItemSelectable,
      suggestionsMenuId,
    } = this.props;

    const items = memoizedItems.map((item, index) => {
      const itemRendered = renderItem(
        item,
        highlightedIndex === index, {
          cursor: 'default'
        }
      );

      const itemIdForRender = `item-${index}`;
      this.references[itemIdForRender] = React.createRef();

      return React.cloneElement(itemRendered, {
        onMouseEnter: isItemSelectable(item) ? () => this.highlightItemFromMouse(index) : null,
        onClick: isItemSelectable(item) ? () => this.selectItemFromMouse(item) : null,
        ref: this.references[itemIdForRender],
        key: itemIdForRender,
        id: `${suggestionsMenuId}-${itemIdForRender}`
      });
    });

    return items;
  }

  renderMenu = (memoizedItems) => {
    const {
      menuLeft,
      menuTop,
      menuWidth,
    } = this.state;

    const {
      renderMenu,
      value,
      suggestionsMenuId,
      menuStyle,
    } = this.props;

    this.references.menu = React.createRef();

    const items = this.renderItem(memoizedItems);

    const style = {
      left: menuLeft,
      top: menuTop,
      minWidth: menuWidth,
    }
    const menu = renderMenu(items, suggestionsMenuId, value, {...menuStyle, ...style})

    return React.cloneElement(menu, {
      ref: this.references.menu,
      // Ignore blur to prevent menu from de-rendering before we can process click
      onTouchStart: () => this.setIgnoreBlur(true),
      onMouseEnter: () => this.setIgnoreBlur(true),
      onMouseLeave: () => this.setIgnoreBlur(false),
    })
  }

  handleInputBlur = (event) => {
    const {
      highlightedIndex
    } = this.state

    const {
      selectOnBlur,
      getItemValue,
      onSelect,
      inputProps,
    } = this.props;

    if (this._ignoreBlur) {
      this._ignoreFocus = true
      this._scrollOffset = getScrollOffset()
     this.references.input.current.focus()
      return
    }
    let setStateCallback

    if (selectOnBlur && highlightedIndex !== null) {
      const items = this.getFilteredItems()
      const item = items[highlightedIndex]
      const value = getItemValue(item)
      setStateCallback = () => onSelect(value, item)
    }

    this.setState({
      isOpen: false,
      highlightedIndex: null
    }, setStateCallback)

    const {
      onBlur
    } = inputProps

    if (onBlur) {
      onBlur(event)
    }
  }

  handleInputFocus = (event) => {
    if (this._ignoreFocus) {
      this._ignoreFocus = false
      const {
        x,
        y
      } = this._scrollOffset
      this._scrollOffset = null
      // Focus will cause the browser to scroll the <input> into view.
      // This can cause the mouse coords to change, which in turn
      // could cause a new highlight to happen, cancelling the click
      // event (when selecting with the mouse)
      window.scrollTo(x, y)
      // Some browsers wait until all focus event handlers have been
      // processed before scrolling the <input> into view, so let's
      // scroll again on the next tick to ensure we're back to where
      // the user was before focus was lost. We could do the deferred
      // scroll only, but that causes a jarring split second jump in
      // some browsers that scroll before the focus event handlers
      // are triggered.
      clearTimeout(this._scrollTimer)
      this._scrollTimer = setTimeout(() => {
        this._scrollTimer = null
        window.scrollTo(x, y)
      }, 0)
      return
    }

    const {
      inputProps,
      openOnFocus,
    } = this.props;

    const {
      onFocus
    } = inputProps

    if (onFocus) {
      if (openOnFocus) {
        this.setState({
          isOpen: true
        })
      }
      onFocus(event)
    }
  }

  isInputFocused = () => {
    const input = this.references.input.current;
    return input.ownerDocument && (input === input.ownerDocument.activeElement)
  }

  handleInputClick = () => {
    const {
      isOpen,
    } = this.state;

    // Input will not be focused if it's disabled
    if (this.isInputFocused() && isOpen)
      this.setState({
        isOpen: true
      })
  }

  composeEventHandlers = (internalEvent, externalEvent) => {
    return externalEvent ? event => { internalEvent(event); externalEvent(event); } : internalEvent
  }

  render() {
    const {
      debug,
      inputProps,
      wrapperStyle,
      wrapperProps,
      renderInput,
      suggestionsMenuId,
      value,
      numberOfResultsAvailableCopy,
      items,
     } = this.props;

     const {
      highlightedIndex,
      isOpen,
     } = this.state;

    const memoizedItems = this.getMemoizedFilteredItems({items, value}) || [];

    if (debug) { // you don't like it, you love it
      this._debugStates.push({
        id: this._debugStates.length,
        state: this.state
      })
    }

    const currentItem = this.references[`item-${highlightedIndex}`];
    const highlightedItemRef = (currentItem && currentItem.current) || null;
    const highlightedItemId = highlightedItemRef ? highlightedItemRef.getAttribute('id'): '';

    const srOnly = {
      clip: 'rect(1px 1px 1px 1px)',
      clip: 'rect(1px, 1px, 1px, 1px)', // eslint-disable-line
      position: 'absolute',
      padding: '0',
      border: '0',
      height: '1px',
      width: '1px',
      overflow: 'hidden',
    };

    return (
      <div style={{ ...wrapperStyle }} {...wrapperProps}>
        <div role="combobox" aria-expanded={isOpen} aria-owns={suggestionsMenuId} aria-haspopup="listbox">
        {renderInput({
          ...inputProps,
          'aria-autocomplete': "list",
          'aria-controls': suggestionsMenuId,
          'aria-activedescendant': highlightedItemId,
          ref: this.references.input,
          onFocus: this.handleInputFocus,
          onBlur: this.handleInputBlur,
          onChange: this.handleChange,
          onKeyDown: this.composeEventHandlers(this.handleKeyDown, inputProps.onKeyDown),
          onClick: this.composeEventHandlers(this.handleInputClick, inputProps.onClick),
          value: value,
        })}
        </div>
        <span style={srOnly} role="status" aria-live="assertive" aria-atomic="true" aria-relevant="aditions">
        {isOpen && value!=='' &&
          `${memoizedItems.length} ${numberOfResultsAvailableCopy}`
        }
        </span>
        {isOpen && this.renderMenu(memoizedItems)}
        {debug && (
          <pre style={{ marginLeft: 300 }}>
            {JSON.stringify(this._debugStates.slice(Math.max(0, this._debugStates.length - 5), this._debugStates.length), null, 2)}
          </pre>
        )}
      </div>
    )
  }
}

module.exports = Autocomplete